File: 288/288676/288676_01.py
#!/usr/bin/env python3
"""b108394.txt から A288676 の b-file を作る。

A288676 は、A108394(k) が素数でないような k を小さい順に並べた数列。
"""

from __future__ import annotations

import argparse
import math
from pathlib import Path
from typing import Iterator


def is_prime(n: int) -> bool:
    """n が素数なら True を返す。"""
    if n < 2:
        return False
    if n % 2 == 0:
        return n == 2

    limit = math.isqrt(n)
    for divisor in range(3, limit + 1, 2):
        if n % divisor == 0:
            return False
    return True


def read_bfile(path: Path) -> Iterator[tuple[int, int]]:
    """b-file の (添字, 値) を順に読み取る。"""
    previous_index: int | None = None

    with path.open(encoding="utf-8-sig") as source:
        for line_number, line in enumerate(source, start=1):
            body = line.partition("#")[0].strip()
            if not body:
                continue

            fields = body.split()
            if len(fields) != 2:
                raise ValueError(
                    f"{path}:{line_number}: 2 列の整数ではありません: {line.rstrip()}"
                )
            try:
                index, value = map(int, fields)
            except ValueError as error:
                raise ValueError(
                    f"{path}:{line_number}: 整数として解釈できません: {line.rstrip()}"
                ) from error

            if previous_index is not None and index <= previous_index:
                raise ValueError(
                    f"{path}:{line_number}: 添字 {index} が昇順になっていません"
                )
            previous_index = index
            yield index, value


def make_a288676(input_path: Path, output_path: Path) -> tuple[int, int, int]:
    """A288676 を計算して出力し、(入力数, 出力数, 最大添字) を返す。"""
    checked = 0
    found = 0
    max_index = 0

    with output_path.open("w", encoding="utf-8", newline="\n") as output:
        output.write(f"# {output_path.name}\n")
        output.write("# A288676: Numbers k such that A108394(k) is not a prime.\n")
        output.write(f"# Computed from {input_path.name}\n")

        for index, value in read_bfile(input_path):
            checked += 1
            max_index = index
            if not is_prime(value):
                found += 1
                output.write(f"{found} {index}\n")

        output.write(f"# a({found + 1}) > {max_index}\n")

    return checked, found, max_index


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="b108394.txt から A288676 の b-file を作成します。"
    )
    parser.add_argument(
        "input",
        nargs="?",
        type=Path,
        default=Path("b108394.txt"),
        help="入力 b-file（既定: b108394.txt）",
    )
    parser.add_argument(
        "-o",
        "--output",
        type=Path,
        default=Path("b288676_01.txt"),
        help="出力 b-file（既定: b288676_01.txt）",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    checked, found, max_index = make_a288676(args.input, args.output)
    print(f"{checked} 項を判定しました（入力の最大添字: {max_index}）")
    print(f"A288676 の {found} 項を {args.output} に出力しました")
    print(f"次の項 a({found + 1}) > {max_index}")


if __name__ == "__main__":
    main()

File: 288/288676/288676_01.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# A288676: Numbers k such that A108394(k) is not a prime.
#
#   A108394 ... k-Goebel 数列 f(0)=1, f(n)=(1+f(0)^n+...+f(n-1)^n)/n が
#               整数でなくなる最初の添字。OEIS でのオフセットは n=2。
#
# 使い方:
#   ruby a288676.rb [b108394.txt] [--list] [--limit N]
#
#   --list     末尾にカンマ区切りの一覧を表示
#   --limit N  b-file の n <= N の行だけを対象にする
#
# 出力: 標準出力に A288676 の b-file 形式 ("1 5", "2 22", ...)
#       進捗や集計は標準エラーへ。

require 'openssl'

# --- 小さい素数（試し割り用） -------------------------------------------
SMALL_PRIMES = begin
  limit = 1000
  sieve = Array.new(limit + 1, true)
  sieve[0] = sieve[1] = false
  (2..Integer.sqrt(limit)).each do |i|
    next unless sieve[i]
    (i * i).step(limit, i) { |j| sieve[j] = false }
  end
  (2..limit).select { |i| sieve[i] }.freeze
end
SMALL_LIMIT = SMALL_PRIMES.last**2  # ここ未満なら試し割りだけで確定

# --- 素数判定 -----------------------------------------------------------
# 小さい数は試し割りで確定、大きい数は OpenSSL の Miller-Rabin。
# (stdlib の 'prime' でもよいが、Prime.prime? は試し割りなので大きい値で遅い)
def prime?(n)
  return false if n < 2

  SMALL_PRIMES.each do |p|
    return true  if n == p
    return false if (n % p).zero?
  end
  return true if n < SMALL_LIMIT

  OpenSSL::BN.new(n).prime?
end

# --- b-file の読み込み --------------------------------------------------
# 形式は "n a(n)"。'#' 以降はコメント、空行は無視。
def each_bfile_entry(path)
  return to_enum(:each_bfile_entry, path) unless block_given?

  File.foreach(path).with_index(1) do |line, lineno|
    body = line.sub(/#.*\z/m, '').strip
    next if body.empty?

    n, a = body.split(/\s+/, 3)
    unless n =~ /\A-?\d+\z/ && a =~ /\A-?\d+\z/
      warn "warning: #{path}:#{lineno}: 解釈できない行をスキップ: #{line.chomp}"
      next
    end
    yield n.to_i, a.to_i
  end
end

# --- main ---------------------------------------------------------------
args  = ARGV.dup
list  = args.delete('--list')
limit = (i = args.index('--limit')) ? args.slice!(i, 2)[1].to_i : nil
path  = args.shift || 'b108394.txt'

abort "#{path} が見つかりません" unless File.exist?(path)

found   = []
checked = 0
maxn    = 0

each_bfile_entry(path) do |n, a|
  next if limit && n > limit

  checked += 1
  maxn = n if n > maxn
  next if prime?(a)

  found << n
  puts "#{found.size} #{n}"
  $stdout.flush
  warn format('  A108394(%d) = %d は素数でない', n, a)
end

warn "---"
warn "#{checked} 項を判定 (n <= #{maxn})、非素数は #{found.size} 個"
warn "次の項 a(#{found.size + 1}) > #{maxn}" if found.any?
puts found.join(', ') if list
